File: visualstudio-extension/src/GitReview.Domain/Porcelain.cs
namespace GitReview.Domain;

public enum ReviewMode
{
    Whole,
    Step,
    Walk,
}

public static class ReviewModeExt
{
    public static string Id(this ReviewMode m) => m switch
    {
        ReviewMode.Whole => "whole",
        ReviewMode.Step => "step",
        ReviewMode.Walk => "walk",
        _ => throw new ArgumentOutOfRangeException(nameof(m)),
    };

    public static ReviewMode? Parse(string? field) => field switch
    {
        "whole" => ReviewMode.Whole,
        "step" => ReviewMode.Step,
        "walk" => ReviewMode.Walk,
        _ => null,
    };
}

public enum WalkthroughStatus
{
    None,
    Applied,
    Degraded,
}

public static class WalkthroughStatusExt
{
    public static string Id(this WalkthroughStatus s) => s switch
    {
        WalkthroughStatus.None => "none",
        WalkthroughStatus.Applied => "applied",
        WalkthroughStatus.Degraded => "degraded",
        _ => throw new ArgumentOutOfRangeException(nameof(s)),
    };

    public static WalkthroughStatus? Parse(string? field) => field switch
    {
        "none" => WalkthroughStatus.None,
        "applied" => WalkthroughStatus.Applied,
        "degraded" => WalkthroughStatus.Degraded,
        _ => null,
    };
}

public sealed record StateRecord(
    string Branch,
    string Source,
    string Tip,
    ReviewMode Mode,
    WalkthroughStatus Walkthrough,
    int? Position = null,
    int? Total = null,
    int? Recorded = null,
    /// <summary>Short SHA (step) or PathRef (walk).</summary>
    object? Current = null,
    bool? Essential = null);

public sealed record EntryRecord(
    int Position,
    /// <summary>Short SHA (step) or PathRef (walk/whole).</summary>
    object Id,
    bool? Essential = null,
    bool? Annotated = null,
    bool? Banked = null);

public sealed record StatusFinishRecord(
    string State = "conflict",
    bool Onto = false);

public sealed record PorcelainResult(
    StateRecord State,
    IReadOnlyList<EntryRecord> Entries,
    IReadOnlyList<EntryRecord>? Files = null,
    StatusFinishRecord? Finish = null,
    bool? Readonly = null,
    bool? KeysOnly = null,
    bool? Draft = null,
    /// <summary>
    /// 012: the absolute path of that draft, as the CLI reported it in the
    /// record's field. Separate from the flag: presence is presence, and a
    /// record without the field (an older CLI) does not turn the mark off.
    /// </summary>
    string? DraftPath = null,
    IReadOnlyDictionary<int, string>? Subjects = null,
    IReadOnlyDictionary<int, string>? Authors = null,
    string? Base = null)
{
    public IReadOnlyList<EntryRecord> FilesList => Files ?? Array.Empty<EntryRecord>();
}

public sealed record BranchFinish(string State, bool Onto);

public sealed record BranchRecord(
    string Name,
    bool Saved,
    bool Current,
    bool Orphan,
    ReviewMode? Mode = null,
    int? Position = null,
    int? Total = null,
    BranchFinish? Finish = null);

public static class Porcelain
{
    private static bool ToBool(string? field) => field == "1";

    private static int ToInt(string? field) =>
        int.TryParse(field, out var n) ? n : 0;

    private static int? ToOptionalInt(string? field)
    {
        if (string.IsNullOrEmpty(field)) return null;
        return int.TryParse(field, out var n) ? n : null;
    }

    /// <summary>
    /// Free-text field: everything after the skip-th tab (subject/author may contain literal tabs).
    /// </summary>
    private static string? RestAfterTab(string line, int skip)
    {
        var index = -1;
        for (var k = 0; k < skip; k++)
        {
            index = line.IndexOf('\t', index + 1);
            if (index == -1) return null;
        }
        return line[(index + 1)..];
    }

    /// <summary>
    /// Tokenizes <c>git review status --porcelain</c>. Mode on the state record is read
    /// first and decides arity of following lines.
    /// </summary>
    public static PorcelainResult ParsePorcelain(string stdout)
    {
        var lines = stdout.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None)
            .Where(l => l.Length > 0)
            .ToList();
        if (lines.Count == 0)
            throw new ArgumentException("porcelain output has no state record");

        StateRecord? state = null;
        var entries = new List<EntryRecord>();
        var files = new List<EntryRecord>();
        Dictionary<int, string>? subjects = null;
        Dictionary<int, string>? authors = null;
        string? bas = null;
        StatusFinishRecord? finish = null;
        bool? isReadonly = null;
        bool? isKeysOnly = null;
        bool? isDraft = null;
        string? draftPath = null;

        foreach (var line in lines)
        {
            var fields = line.Split('\t');
            switch (fields[0])
            {
                case "state":
                {
                    var mode = ReviewModeExt.Parse(Get(fields, 4))
                        ?? throw new ArgumentException(
                            $"porcelain state has invalid mode: {(Get(fields, 4) is { } m ? $"\"{m}\"" : "(missing)")}");
                    var walkthrough = WalkthroughStatusExt.Parse(Get(fields, 5))
                                      ?? WalkthroughStatus.None;
                    StateRecord record;
                    if (mode is ReviewMode.Step or ReviewMode.Walk)
                    {
                        object? current = mode == ReviewMode.Walk
                            ? Unquote.ToPathRef(Get(fields, 9) ?? "")
                            : Get(fields, 9);
                        record = new StateRecord(
                            Branch: Get(fields, 1) ?? "",
                            Source: Get(fields, 2) ?? "",
                            Tip: Get(fields, 3) ?? "",
                            Mode: mode,
                            Walkthrough: walkthrough,
                            Position: ToInt(Get(fields, 6)),
                            Total: ToInt(Get(fields, 7)),
                            Recorded: ToInt(Get(fields, 8)),
                            Current: current,
                            Essential: mode == ReviewMode.Walk ? ToBool(Get(fields, 10)) : null);
                    }
                    else
                    {
                        record = new StateRecord(
                            Branch: Get(fields, 1) ?? "",
                            Source: Get(fields, 2) ?? "",
                            Tip: Get(fields, 3) ?? "",
                            Mode: mode,
                            Walkthrough: walkthrough);
                    }
                    state = record;
                    break;
                }
                case "entry":
                {
                    var st = state ?? throw new ArgumentException("entry record before state record");
                    var position = ToInt(Get(fields, 1));
                    var rawId = Get(fields, 2) ?? "";
                    object id = st.Mode == ReviewMode.Step ? rawId : Unquote.ToPathRef(rawId);
                    EntryRecord entry = st.Mode switch
                    {
                        ReviewMode.Walk => new EntryRecord(
                            position, id,
                            Essential: ToBool(Get(fields, 3)),
                            Annotated: ToBool(Get(fields, 4))),
                        ReviewMode.Step => new EntryRecord(
                            position, id,
                            Banked: ToBool(Get(fields, 3))),
                        _ => new EntryRecord(position, id),
                    };
                    entries.Add(entry);
                    break;
                }
                case "file":
                {
                    if (state is null) throw new ArgumentException("file record before state record");
                    var position = ToInt(Get(fields, 1));
                    var rawPath = Get(fields, 2);
                    if (string.IsNullOrEmpty(rawPath)) continue;
                    files.Add(new EntryRecord(position, Unquote.ToPathRef(rawPath)));
                    break;
                }
                case "subject":
                case "author":
                {
                    var position = ToOptionalInt(Get(fields, 1));
                    var text = RestAfterTab(line, 2);
                    if (position is null || text is null) continue;
                    if (fields[0] == "subject")
                    {
                        subjects ??= new Dictionary<int, string>();
                        subjects[position.Value] = text;
                    }
                    else
                    {
                        authors ??= new Dictionary<int, string>();
                        authors[position.Value] = text;
                    }
                    break;
                }
                case "base":
                {
                    var text = RestAfterTab(line, 1);
                    if (text is not null) bas = text;
                    break;
                }
                case "finish":
                {
                    if (Get(fields, 1) == "conflict")
                        finish = new StatusFinishRecord(Onto: ToBool(Get(fields, 2)));
                    break;
                }
                case "readonly":
                    isReadonly = true;
                    break;
                case "keys":
                    isKeysOnly = true;
                    break;
                case "draft":
                    isDraft = true;
                    if (!string.IsNullOrEmpty(Get(fields, 1))) draftPath = Get(fields, 1);
                    break;
                // unknown tag: ignore (FR-003)
            }
        }

        var stFinal = state ?? throw new ArgumentException("porcelain output has no state record");
        return new PorcelainResult(
            State: stFinal,
            Entries: entries,
            Files: files,
            Finish: finish,
            Readonly: isReadonly,
            KeysOnly: isKeysOnly,
            Draft: isDraft,
            DraftPath: draftPath,
            Subjects: subjects,
            Authors: authors,
            Base: bas);
    }

    /// <summary>Source name of a review branch: strips review-saved/ or review/ prefix.</summary>
    public static string SourceOf(BranchRecord branch)
    {
        foreach (var prefix in new[] { "review-saved/", "review/" })
        {
            if (branch.Name.StartsWith(prefix, StringComparison.Ordinal))
                return branch.Name[prefix.Length..];
        }
        return branch.Name;
    }

    /// <summary>Tokenizes <c>git review list --porcelain</c>. Empty output is valid.</summary>
    public static IReadOnlyList<BranchRecord> ParseListPorcelain(string stdout)
    {
        var branches = new List<BranchRecord>();
        var finishByBranch = new Dictionary<string, BranchFinish>();

        foreach (var line in stdout.Split(new[] { "\r\n", "\n" }, StringSplitOptions.None))
        {
            if (line.Length == 0) continue;
            var fields = line.Split('\t');
            switch (fields[0])
            {
                case "finish":
                {
                    var branchName = Get(fields, 1);
                    var state = Get(fields, 2);
                    if (branchName is not null && (state == "pending" || state == "conflict"))
                        finishByBranch[branchName] = new BranchFinish(state, ToBool(Get(fields, 3)));
                    break;
                }
                case "branch":
                {
                    var orphan = ToBool(Get(fields, 4));
                    int? pos = null;
                    int? tot = null;
                    ReviewMode? mode = null;
                    if (!orphan)
                    {
                        mode = ReviewModeExt.Parse(Get(fields, 5) ?? "whole");
                        var p = ToOptionalInt(Get(fields, 6));
                        var t = ToOptionalInt(Get(fields, 7));
                        if (p is not null && t is not null)
                        {
                            pos = p;
                            tot = t;
                        }
                    }
                    branches.Add(new BranchRecord(
                        Name: Get(fields, 1) ?? "",
                        Saved: ToBool(Get(fields, 2)),
                        Current: ToBool(Get(fields, 3)),
                        Orphan: orphan,
                        Mode: mode,
                        Position: pos,
                        Total: tot));
                    break;
                }
            }
        }

        return branches.Select(b =>
            finishByBranch.TryGetValue(b.Name, out var f) ? b with { Finish = f } : b
        ).ToList();
    }

    private static string? Get(string[] fields, int i) =>
        i < fields.Length ? fields[i] : null;
}

File: jetbrains-plugin/src/main/kotlin/com/ezevillo/gitreview/domain/Porcelain.kt
package com.ezevillo.gitreview.domain

enum class ReviewMode {
    WHOLE, STEP, WALK;

    val id: String
        get() = when (this) {
            WHOLE -> "whole"
            STEP -> "step"
            WALK -> "walk"
        }

    companion object {
        fun parse(field: String?): ReviewMode? = when (field) {
            "whole" -> WHOLE
            "step" -> STEP
            "walk" -> WALK
            else -> null
        }
    }
}

enum class WalkthroughStatus {
    NONE, APPLIED, DEGRADED;

    val id: String
        get() = when (this) {
            NONE -> "none"
            APPLIED -> "applied"
            DEGRADED -> "degraded"
        }

    companion object {
        fun parse(field: String?): WalkthroughStatus? = when (field) {
            "none" -> NONE
            "applied" -> APPLIED
            "degraded" -> DEGRADED
            else -> null
        }
    }
}

data class StateRecord(
    val branch: String,
    val source: String,
    val tip: String,
    val mode: ReviewMode,
    val walkthrough: WalkthroughStatus,
    val position: Int? = null,
    val total: Int? = null,
    val recorded: Int? = null,
    /** Short SHA (step) or [PathRef] (walk). */
    val current: Any? = null,
    val essential: Boolean? = null,
)

data class EntryRecord(
    val position: Int,
    /** Short SHA (step) or [PathRef] (walk/whole). */
    val id: Any,
    val essential: Boolean? = null,
    val annotated: Boolean? = null,
    val banked: Boolean? = null,
)

data class StatusFinishRecord(
    val state: String = "conflict",
    val onto: Boolean,
)

data class PorcelainResult(
    val state: StateRecord,
    val entries: List<EntryRecord>,
    /**
     * Step-mode inventory of files in the *current* commit (`file` records).
     * Position is 1-based within that commit. Empty when the CLI emits none
     * (empty commit, or walk/whole where `file` is not used).
     */
    val files: List<EntryRecord> = emptyList(),
    val finish: StatusFinishRecord? = null,
    val readonly: Boolean? = null,
    val keysOnly: Boolean? = null,
    /**
     * 011: el orden de lectura es el borrador del revisor, no el walkthrough
     * del autor. Sólo se da en walk; nunca se infiere, llega por el registro.
     */
    val draft: Boolean? = null,
    /**
     * 012: la ruta absoluta de ese borrador, tal como la reportó la CLI en el
     * campo del registro. Aparte del booleano: la presencia sigue siendo la
     * presencia, y un registro sin campo (una CLI anterior) no la apaga.
     */
    val draftPath: String? = null,
    val subjects: Map<Int, String>? = null,
    val authors: Map<Int, String>? = null,
    val base: String? = null,
)

data class BranchFinish(
    val state: String, // "pending" | "conflict"
    val onto: Boolean,
)

data class BranchRecord(
    val name: String,
    val saved: Boolean,
    val current: Boolean,
    val orphan: Boolean,
    val mode: ReviewMode? = null,
    val position: Int? = null,
    val total: Int? = null,
    val finish: BranchFinish? = null,
)

/**
 * How much work dropping a `review-fixes/` branch would cost.
 *
 * `empty` is not a flavour of "safe": an untouched fixes branch sits at the PR
 * tip and holds NOTHING of yours, which is a different thing from being already
 * integrated. `unknown` does not fold into `unmerged` either -- with no base
 * configured the question has no answer, and giving the worse of the two paints
 * a branch that may be empty as dangerous.
 */
enum class FixesState {
    EMPTY,
    MERGED,
    UNMERGED,
    UNKNOWN,
    ;

    companion object {
        fun parse(field: String?): FixesState = when (field) {
            "empty" -> EMPTY
            "merged" -> MERGED
            "unmerged" -> UNMERGED
            // Anything we do not understand reads as "cannot tell", never as one
            // of the three: this badge is the only thing between dropping an
            // empty branch and dropping unpushed work.
            else -> UNKNOWN
        }
    }
}

/** A `fixes` record of `git review list --porcelain`: one `review-fixes/` branch. */
data class FixesRecord(
    val name: String,
    val current: Boolean,
    val session: Boolean,
    val state: FixesState,
)

private fun toBool(field: String?): Boolean = field == "1"

private fun toInt(field: String?): Int = field?.toIntOrNull() ?: 0

private fun toOptionalInt(field: String?): Int? {
    if (field.isNullOrEmpty()) return null
    return field.toIntOrNull()
}

/**
 * Free-text field: everything after the `skip`-th tab (subject/author may
 * contain literal tabs).
 */
private fun restAfterTab(line: String, skip: Int): String? {
    var index = -1
    repeat(skip) {
        index = line.indexOf('\t', index + 1)
        if (index == -1) return null
    }
    return line.substring(index + 1)
}

/**
 * Tokenizes `git review status --porcelain`. Mode on the state record is read
 * first and decides arity of following lines.
 */
fun parsePorcelain(stdout: String): PorcelainResult {
    val lines = stdout.split(Regex("\r?\n")).filter { it.isNotEmpty() }
    if (lines.isEmpty()) {
        throw IllegalArgumentException("porcelain output has no state record")
    }

    var state: StateRecord? = null
    val entries = ArrayList<EntryRecord>()
    val files = ArrayList<EntryRecord>()
    var subjects: MutableMap<Int, String>? = null
    var authors: MutableMap<Int, String>? = null
    var base: String? = null
    var finish: StatusFinishRecord? = null
    var isReadonly: Boolean? = null
    var isKeysOnly: Boolean? = null
    var isDraft: Boolean? = null
    var draftPath: String? = null

    for (line in lines) {
        val fields = line.split("\t")
        when (fields[0]) {
            "state" -> {
                val mode = ReviewMode.parse(fields.getOrNull(4))
                    ?: throw IllegalArgumentException(
                        "porcelain state has invalid mode: ${fields.getOrNull(4)?.let { "\"$it\"" } ?: "(missing)"}",
                    )
                val walkthrough = WalkthroughStatus.parse(fields.getOrNull(5))
                    ?: WalkthroughStatus.NONE
                val record = if (mode == ReviewMode.STEP || mode == ReviewMode.WALK) {
                    val current: Any? = if (mode == ReviewMode.WALK) {
                        toPathRef(fields.getOrNull(9) ?: "")
                    } else {
                        fields.getOrNull(9)
                    }
                    StateRecord(
                        branch = fields.getOrNull(1) ?: "",
                        source = fields.getOrNull(2) ?: "",
                        tip = fields.getOrNull(3) ?: "",
                        mode = mode,
                        walkthrough = walkthrough,
                        position = toInt(fields.getOrNull(6)),
                        total = toInt(fields.getOrNull(7)),
                        recorded = toInt(fields.getOrNull(8)),
                        current = current,
                        essential = if (mode == ReviewMode.WALK) toBool(fields.getOrNull(10)) else null,
                    )
                } else {
                    StateRecord(
                        branch = fields.getOrNull(1) ?: "",
                        source = fields.getOrNull(2) ?: "",
                        tip = fields.getOrNull(3) ?: "",
                        mode = mode,
                        walkthrough = walkthrough,
                    )
                }
                state = record
            }
            "entry" -> {
                val st = state ?: throw IllegalArgumentException("entry record before state record")
                val position = toInt(fields.getOrNull(1))
                val rawId = fields.getOrNull(2) ?: ""
                val id: Any = if (st.mode == ReviewMode.STEP) rawId else toPathRef(rawId)
                val entry = when (st.mode) {
                    ReviewMode.WALK -> EntryRecord(
                        position = position,
                        id = id,
                        essential = toBool(fields.getOrNull(3)),
                        annotated = toBool(fields.getOrNull(4)),
                    )
                    ReviewMode.STEP -> EntryRecord(
                        position = position,
                        id = id,
                        banked = toBool(fields.getOrNull(3)),
                    )
                    ReviewMode.WHOLE -> EntryRecord(position = position, id = id)
                }
                entries.add(entry)
            }
            // Step: files of the current commit only (path as PathRef, like whole).
            "file" -> {
                state ?: throw IllegalArgumentException("file record before state record")
                val position = toInt(fields.getOrNull(1))
                val rawPath = fields.getOrNull(2)
                if (rawPath.isNullOrEmpty()) continue
                files.add(EntryRecord(position = position, id = toPathRef(rawPath)))
            }
            "subject", "author" -> {
                val position = toOptionalInt(fields.getOrNull(1))
                val text = restAfterTab(line, 2)
                if (position == null || text == null) continue
                if (fields[0] == "subject") {
                    subjects = (subjects ?: mutableMapOf()).also { it[position] = text }
                } else {
                    authors = (authors ?: mutableMapOf()).also { it[position] = text }
                }
            }
            "base" -> {
                val text = restAfterTab(line, 1)
                if (text != null) base = text
            }
            "finish" -> {
                if (fields.getOrNull(1) == "conflict") {
                    finish = StatusFinishRecord(onto = toBool(fields.getOrNull(2)))
                }
            }
            "readonly" -> isReadonly = true
            "keys" -> isKeysOnly = true
            "draft" -> {
                isDraft = true
                fields.getOrNull(1)?.takeIf { it.isNotEmpty() }?.let { draftPath = it }
            }
            else -> { /* unknown tag: ignore (FR-003) */ }
        }
    }

    val st = state ?: throw IllegalArgumentException("porcelain output has no state record")
    return PorcelainResult(
        state = st,
        entries = entries,
        files = files,
        finish = finish,
        readonly = isReadonly,
        keysOnly = isKeysOnly,
        draft = isDraft,
        draftPath = draftPath,
        subjects = subjects,
        authors = authors,
        base = base,
    )
}

/**
 * Source name of a review branch: strips `review-saved/` or `review/` prefix.
 */
fun sourceOf(branch: BranchRecord): String {
    for (prefix in listOf("review-saved/", "review/")) {
        if (branch.name.startsWith(prefix)) {
            return branch.name.removePrefix(prefix)
        }
    }
    return branch.name
}

/**
 * The `fixes` records of the same output, in a function of their own rather than
 * a second return value of [parseListPorcelain]: these are branches of *edits*,
 * not reviews -- there is nothing to resume or abort on them -- and every
 * existing consumer of the inventory keeps asking for exactly what it asked for.
 */
fun parseListFixes(stdout: String): List<FixesRecord> {
    val fixes = ArrayList<FixesRecord>()
    for (line in stdout.split(Regex("\r?\n"))) {
        if (line.isEmpty()) continue
        val fields = line.split("\t")
        if (fields[0] != "fixes") continue
        val name = fields.getOrNull(1)
        if (name.isNullOrEmpty()) continue
        fixes.add(
            FixesRecord(
                name = name,
                current = toBool(fields.getOrNull(2)),
                session = toBool(fields.getOrNull(3)),
                state = FixesState.parse(fields.getOrNull(4)),
            )
        )
    }
    return fixes
}

/** Tokenizes `git review list --porcelain`. Empty output is valid. */
fun parseListPorcelain(stdout: String): List<BranchRecord> {
    val branches = ArrayList<BranchRecord>()
    val finishByBranch = HashMap<String, BranchFinish>()

    for (line in stdout.split(Regex("\r?\n"))) {
        if (line.isEmpty()) continue
        val fields = line.split("\t")
        when (fields[0]) {
            "finish" -> {
                val branchName = fields.getOrNull(1)
                val state = fields.getOrNull(2)
                if (branchName != null && (state == "pending" || state == "conflict")) {
                    finishByBranch[branchName] = BranchFinish(state = state, onto = toBool(fields.getOrNull(3)))
                }
            }
            "branch" -> {
                val record = BranchRecord(
                    name = fields.getOrNull(1) ?: "",
                    saved = toBool(fields.getOrNull(2)),
                    current = toBool(fields.getOrNull(3)),
                    orphan = toBool(fields.getOrNull(4)),
                    mode = if (!toBool(fields.getOrNull(4))) {
                        ReviewMode.parse(fields.getOrNull(5) ?: "whole")
                    } else null,
                    position = if (!toBool(fields.getOrNull(4))) {
                        val p = toOptionalInt(fields.getOrNull(6))
                        val t = toOptionalInt(fields.getOrNull(7))
                        if (p != null && t != null) p else null
                    } else null,
                    total = if (!toBool(fields.getOrNull(4))) {
                        val p = toOptionalInt(fields.getOrNull(6))
                        val t = toOptionalInt(fields.getOrNull(7))
                        if (p != null && t != null) t else null
                    } else null,
                )
                branches.add(record)
            }
        }
    }

    return branches.map { b ->
        val f = finishByBranch[b.name]
        if (f != null) b.copy(finish = f) else b
    }
}
